torch/lib/c10d/test/ProcessGroupNCCLTest.cpp

```diff
@@ -386,6 +403,45 @@ void testAllgather(const std::string& path, int rank, int size) {
   std::cout << "Allgather test successful" << std::endl;
 }
 
+void testAllgatherWithTimeout(const std::string& path, int rank, int size) {
+  // First we must set NCCL_BLOCKING_WAIT to 1 for timeouts to work in
+  // ProcessGroupNCCL.
+  auto originalBlockingWait = getenv(c10d::NCCL_BLOCKING_WAIT);
+  setenv(c10d::NCCL_BLOCKING_WAIT, "1", 1);
+  auto test = AllgatherNCCLTestWithTimeout(path, size);
+  test.initialize(rank, size);
+  auto work = test.run();
+
+  // Wait for work to finish
+  try {
+    test.wait(work, kWorkTimeout);
+  } catch (const std::runtime_error& e) {
+    // First we reset the NCCL_BLOCKING_WAIT environment variable to prevent
+    // any unwanted side-effects. We must do this at all exit-points for this
+    // function.
+    setenv(c10d::NCCL_BLOCKING_WAIT, originalBlockingWait, 1);
+    // We are expecting the operation to timeout and throw a runtime_error
+    std::string errorMsg = e.what();
+    if (errorMsg.find(kTimeoutErrorString) != std::string::npos) {
+      // If the runtime_error message is the same one thrown in wait::timeout,
+      // the test is successful
+      std::cout << "Allgather test with timeouts was successful" << std::endl;
+      return;
+    } else {
+      // Other runtime errors indicates a test failure
+      throw std::runtime_error("BOOM!");
+    }
+  } catch (...) {
+    setenv(c10d::NCCL_BLOCKING_WAIT, originalBlockingWait, 1);
+    // Any other exception indicates a test failure
+    throw std::runtime_error("BOOM!");
+  }
+  setenv(c10d::NCCL_BLOCKING_WAIT, originalBlockingWait, 1);
+  // If no exception, that is also an error since we expect the test.wait call
+  // to timeout and throw.
+  throw std::runtime_error("BOOM!");
```

Comment:
Can we make the message more informative here? Same for the above two throws.

Reply:
I was actually planning on cleaning up these messages, but just followed the same error messages as the already existing tests for now (they all do `throw std::runtime_error("BOOM!")` for some reason...). I'm planning on putting up another PR that fixes all of these.